storm-client/src/jvm/org/apache/storm/windowing/TimeTriggerPolicy.java

```diff
@@ -109,13 +112,22 @@ private Runnable newTriggerTask() {
         return new Runnable() {
             @Override
             public void run() {
+                // compute the current trigger timestamp based on prevTriggerTimestamp,
+                // since the calculation based on System.currentTimeMillis might have a slight drift
+                long now = System.currentTimeMillis();
+                long triggerTs = prevTriggerTimestamp == 0 ? now : prevTriggerTimestamp + duration;
+                prevTriggerTimestamp = triggerTs;
+                if (Math.abs(triggerTs - now) > 1000) {
```

Comment:
The timing issues are due to the scheduled executor (which uses nanoTime) which does not exactly trigger after the duration when measured with System.currentTimeMillis. Theres a mismatch of a few ms, which is addressed here. In general user code is expected to return before the next trigger. If user code takes more time than the trigger interval still we set the next trigger timestamp correctly and execute as soon as they return from execute, but this should be very rare and users are not expected to do this, so we log the warning. I don't think we want multiple concurrent executes (i.e trigger execute when they haven't returned from the first execute)